crates/trie/trie/src/proof_v2/mod.rs

```diff
@@ -570,27 +646,450 @@ where
+            branch=?self.branch_stack.last(),
+            branch_path=?self.branch_path,
+            "Pushed cached branch",
+        );
+
+        Ok(())
+    }
+
+    /// Accepts the current state of both hashed and trie cursors, and determines the next range of
+    /// hashed keys which need to be processed using [`Self::push_leaf`].
+    ///
+    /// This method will use cached branch node data from the trie cursor to skip over all possible
+    /// ranges of keys, to reduce computation as much as possible.
+    ///
+    /// # Returns
+    ///
+    /// - `None`: No more data to process, finish computation
+    ///
+    /// - `Some(lower, None)`: Indicates to call `push_leaf` on all keys starting at `lower`, with
+    ///   no upper bound. This method won't be called again after this.
+    ///
+    /// - `Some(lower, Some(upper))`: Indicates to call `push_leaf` on all keys starting at `lower`,
+    ///   up to but excluding `upper`, and then call this method once done.
+    #[instrument(target = TRACE_TARGET, level = "trace", skip_all)]
+    fn next_uncached_key_range(
+        &mut self,
+        targets: &mut TargetsIter<impl Iterator<Item = Nibbles>>,
+        trie_cursor_state: &mut TrieCursorState,
+        hashed_key_current_path: Option<Nibbles>,
+    ) -> Result<Option<(Nibbles, Option<Nibbles>)>, StateProofError> {
+        // All trie data prior to the current cached branch, if any, has been computed. Any branches
+        // which were under-construction previously, and which are not on the same path as this
+        // cached branch, can be assumed to be completed; they will not have any further keys added
+        // to them.
+        if let Some(cached_path) = self.cached_branch_stack.last().map(|kv| kv.0) {
+            while !cached_path.starts_with(&self.branch_path) {
+                self.pop_branch(targets)?;
+            }
+        }
+
+        // `lower_bound` will be used to track the lower bound of the range which is returned from
+        // this method. If this is None then there are no further keys which need to be processed.
+        //
+        // This starts off being based off of the hashed cursor's current position, which is the
+        // next key which hasn't been processed. If that is None then we start from zero.
+        let mut lower_bound = Some(hashed_key_current_path.unwrap_or_default());
+
+        loop {
+            // Determine the current cached branch node.
+            //
+            // NOTE we pop off the `cached_branch_stack` because cloning the `BranchNodeCompact`
+            // means cloning an Arc, which incurs synchronization overhead. We have to be sure to
+            // push the cached branch back onto the stack once done.
+            let (cached_path, cached_branch) = match (
+                self.cached_branch_stack.pop(),
+                &trie_cursor_state,
+                lower_bound.as_ref(),
+            ) {
+                (Some(cached), _, _) => cached,
```

Comment:
i find these match condition slightly hard to follow as it requires some sort of context understanding on the possible states. 

e.g 
```
// Use Queued Cache, cached_branch_stack has something
(Some(cached), _, _) => cached

// No More Cache Data
(None, TrieCursorState::Exhausted, _) | (_, _, None) 

```

had to generate an ascii digram to understand the flow better
https://gist.github.com/yongkangc/1e571093db035be6e09942b8eb93cbfe

would a helper method with enum be much better?